setup.sh

```diff
@@ -17,35 +17,19 @@ function check_installed_python() {
         exit 2
     fi
 
-    which python3.8
-    if [ $? -eq 0 ]; then
-        echo "using Python 3.8"
-        PYTHON=python3.8
-        check_installed_pip
-        return
-    fi
-
-    which python3.9
-    if [ $? -eq 0 ]; then
-        echo "using Python 3.9"
-        PYTHON=python3.9
-        check_installed_pip
-        return
-    fi
-
-    which python3.7
-    if [ $? -eq 0 ]; then
-        echo "using Python 3.7"
-        PYTHON=python3.7
-        check_installed_pip
-        return
-    fi
+    for v in {9..7}; do
+        PYTHON="python3.${v}"
+        which $PYTHON
+        if [ $? -eq 0 ]; then
+            echo "using ${PYTHON}"
```

Comment:
To be honest, it's intentional to check for python 3.8 first.

Python 3.9 (despite now being out for quite some time) still has some problems with non-available wheels - which means that with python 3.8 (if available) - your install chances are better by using that.

e.g. if you use 3.8, you'll not need blosc / hdf5 dependencies (not just on mac) - which will be needed for 3.9.
As such - while this will simplify the method seemingly - it'll change behaviour in an unintended way.

Reply:
I'll switch it back for now, 3.8 might cause problems on mac, it's working for me right now but I might have some extra things installed.

I think 3.8 might only work for certain operating systems and setups

Both `3.8` and `3.9` fail on ubuntu20&18, I just started a brand new ec2, ran these commands to get a little bit further, but then received the problems described below

```
   git clone https://github.com/freqtrade/freqtrade
   curl https://bootstrap.pypa.io/get-pip.py -o get-pip.py
   
   python3.8 get-pip.py
   python3.8 -m install ensurepip
   sudo apt install python3.8-venv
   
   sudo apt-get install python3.9
   python3.9 get-pip.py 
   python3.9 -m pip install ensurepip
   sudo apt install python3.9-venv
```

For 3.8, this line runs forever
```
    Running setup.py install for TA-Lib ... /
```

For 3.9, I get this
```
ERROR: Could not build wheels for py-find-1st which use PEP 517 and cannot be installed directly
```

I'm able to get 3.7 working on ubuntu if I run this script

```
curPath=${pwd}
sudo apt-get update
sudo apt-get install gcc
# sudo apt install wget
sudo apt-get install libreadline-gplv2-dev libncursesw5-dev libssl-dev libsqlite3-dev tk-dev libgdbm-dev libc6-dev libbz2-dev libffi-dev wget liblzma-dev lzma
cd /usr/src
sudo wget https://www.python.org/ftp/python/3.7.9/Python-3.7.9.tgz
sudo tar xzf Python-3.7.9.tgz
cd Python-3.7.9
sudo ./configure --enable-optimizations
sudo make
sudo make altinstall
```

---------------------

**ec2 info**

Ubuntu 20.04.2 LTS
Python 3.9.5
Python 3.8.10

Reply:
> Running setup.py install for TA-Lib ... /

Just be patient. It's compiling ta-lib ... and might take a while. it'll not run "forever" - but might run for a long time if you use a low-end machine - or timeout/crash if it's not got enough memory.

Reply:
> > Running setup.py install for TA-Lib ... /
> 
> Just be patient. It's compiling ta-lib ... and might take a while. it'll not run "forever" - but might run for a long time if you use a low-end machine - or timeout/crash if it's not got enough memory.

I think I've let it run for an hour before on those machines and it hasn't finished, I'm not sure because this was a few months ago

Reply:
@xmatthias I'm trying to run the setup.sh script on an Ubuntu20 ec2 (t2.micro) right now, it's been stuck on `Running setup.py install for TA-Lib ... |` for the last 45 minutes, I can let it keep running and see if it ever finishes but I think it's stuck. It does only have 1 vCPU though. I can install it when using python3.7 though

I guess mine might be niche case, but I usually use these kind of ec2's for dry-runs and live, I don't know if other's do.

Reply:
a t2.micro will not fullfill the minimum requirements we list on the readme.
While that does not prevent you from running the bot - it'll usually prevent you from installing it natively (that's where docker really shines).

relying on python3.7 is a dangerous approach in general - as it's now 2 major releases behind - which means dependencies might drop support for it at any time (with which freqtrade will follow suit almost immediately - as 3.8 and 3.9 are out for long enough).
Also, if we come across a feature that prevents us from supporting 3.7 - support for that will be dropped, too.

If installation for it takes a long time for ta-lib - you might wanna check out the ta-lib repository if they do something "special" for 3.7 (you can do many things in setup.py files) ...

Reply:
I'll check out if I can get 3.9 working, 3.8 eventually failed

```
 Running setup.py install for TA-Lib ... error
    ERROR: Command errored out with exit status 1:
     command: /home/ubuntu/freqtrade/.env/bin/python3.8 -u -c 'import io, os, sys, setuptools, tokenize; sys.argv[0] = '"'"'/tmp/pip-install-pi3sc42f/ta-lib_a20f0cc09394411686db806a1834e6eb/setup.py'"'"'; __file__='"'"'/tmp/pip-install-pi3sc42f/ta-lib_a20f0cc09394411686db806a1834e6eb/setup.py'"'"';f = getattr(tokenize, '"'"'open'"'"', open)(__file__) if os.path.exists(__file__) else io.StringIO('"'"'from setuptools import setup; setup()'"'"');code = f.read().replace('"'"'\r\n'"'"', '"'"'\n'"'"');f.close();exec(compile(code, __file__, '"'"'exec'"'"'))' install --record /tmp/pip-record-k84bu4gk/install-record.txt --single-version-externally-managed --compile --install-headers /home/ubuntu/freqtrade/.env/include/site/python3.8/TA-Lib
         cwd: /tmp/pip-install-pi3sc42f/ta-lib_a20f0cc09394411686db806a1834e6eb/
    Complete output (80 lines):
    running install
    running build
    running build_py
    creating build
    creating build/lib.linux-x86_64-3.8
    creating build/lib.linux-x86_64-3.8/talib
    copying talib/test_pandas.py -> build/lib.linux-x86_64-3.8/talib
    copying talib/test_func.py -> build/lib.linux-x86_64-3.8/talib
    copying talib/abstract.py -> build/lib.linux-x86_64-3.8/talib
    copying talib/deprecated.py -> build/lib.linux-x86_64-3.8/talib
    copying talib/stream.py -> build/lib.linux-x86_64-3.8/talib
    copying talib/test_abstract.py -> build/lib.linux-x86_64-3.8/talib
    copying talib/__init__.py -> build/lib.linux-x86_64-3.8/talib
    copying talib/test_data.py -> build/lib.linux-x86_64-3.8/talib
    copying talib/test_stream.py -> build/lib.linux-x86_64-3.8/talib
    running build_ext
    building 'talib._ta_lib' extension
    creating build/temp.linux-x86_64-3.8
    creating build/temp.linux-x86_64-3.8/talib
    x86_64-linux-gnu-gcc -pthread -Wno-unused-result -Wsign-compare -DNDEBUG -g -fwrapv -O2 -Wall -g -fstack-protector-strong -Wformat -Werror=format-security -g -fwrapv -O2 -g -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -fPIC -I/usr/include -I/usr/local/include -I/opt/include -I/opt/local/include -I/opt/homebrew/include -I/opt/homebrew/opt/ta-lib/include -I/home/ubuntu/freqtrade/.env/lib/python3.8/site-packages/numpy/core/include -I/home/ubuntu/freqtrade/.env/include -I/usr/include/python3.8 -c talib/_ta_lib.c -o build/temp.linux-x86_64-3.8/talib/_ta_lib.o
    In file included from /home/ubuntu/freqtrade/.env/lib/python3.8/site-packages/numpy/core/include/numpy/ndarraytypes.h:1969,
                     from /home/ubuntu/freqtrade/.env/lib/python3.8/site-packages/numpy/core/include/numpy/ndarrayobject.h:12,
                     from /home/ubuntu/freqtrade/.env/lib/python3.8/site-packages/numpy/core/include/numpy/arrayobject.h:4,
                     from talib/_ta_lib.c:620:
    /home/ubuntu/freqtrade/.env/lib/python3.8/site-packages/numpy/core/include/numpy/npy_1_7_deprecated_api.h:17:2: warning: #warning "Using deprecated NumPy API, disable it with " "#define NPY_NO_DEPRECATED_API NPY_1_7_API_VERSION" [-Wcpp]
       17 | #warning "Using deprecated NumPy API, disable it with " \
          |  ^~~~~~~
    talib/_ta_lib.c: In function ‘__pyx_pf_5talib_7_ta_lib_348_ta_getFuncInfo’:
    talib/_ta_lib.c:38240:104: warning: passing argument 2 of ‘TA_GetFuncInfo’ from incompatible pointer type [-Wincompatible-pointer-types]
    38240 |   __pyx_v_retCode = TA_GetFuncInfo(__pyx_f_5talib_7_ta_lib___ta_getFuncHandle(__pyx_v_function_name), (&__pyx_v_info));
          |                                                                                                       ~^~~~~~~~~~~~~~
          |                                                                                                        |
          |                                                                                                        TA_FuncInfo ** {aka struct TA_FuncInfo **}
    In file included from talib/_ta_lib.c:615:
    /usr/local/include/ta-lib/ta_abstract.h:211:48: note: expected ‘const TA_FuncInfo **’ {aka ‘const struct TA_FuncInfo **’} but argument is of type ‘TA_FuncInfo **’ {aka ‘struct TA_FuncInfo **’}
      211 |                            const TA_FuncInfo **funcInfo );
          |                            ~~~~~~~~~~~~~~~~~~~~^~~~~~~~
    talib/_ta_lib.c: In function ‘__pyx_pf_5talib_7_ta_lib_350_ta_getInputParameterInfo’:
    talib/_ta_lib.c:38511:127: warning: passing argument 3 of ‘TA_GetInputParameterInfo’ from incompatible pointer type [-Wincompatible-pointer-types]
    38511 |   __pyx_v_retCode = TA_GetInputParameterInfo(__pyx_f_5talib_7_ta_lib___ta_getFuncHandle(__pyx_v_function_name), __pyx_v_idx, (&__pyx_v_info));
          |                                                                                                                              ~^~~~~~~~~~~~~~
          |                                                                                                                               |
          |                                                                                                                               TA_InputParameterInfo ** {aka struct TA_InputParameterInfo **}
    In file included from talib/_ta_lib.c:615:
    /usr/local/include/ta-lib/ta_abstract.h:442:68: note: expected ‘const TA_InputParameterInfo **’ {aka ‘const struct TA_InputParameterInfo **’} but argument is of type ‘TA_InputParameterInfo **’ {aka ‘struct TA_InputParameterInfo **’}
      442 |                                      const TA_InputParameterInfo **info );
          |                                      ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~
    talib/_ta_lib.c: In function ‘__pyx_pf_5talib_7_ta_lib_352_ta_getOptInputParameterInfo’:
    talib/_ta_lib.c:38761:130: warning: passing argument 3 of ‘TA_GetOptInputParameterInfo’ from incompatible pointer type [-Wincompatible-pointer-types]
    38761 |   __pyx_v_retCode = TA_GetOptInputParameterInfo(__pyx_f_5talib_7_ta_lib___ta_getFuncHandle(__pyx_v_function_name), __pyx_v_idx, (&__pyx_v_info));
          |                                                                                                                                 ~^~~~~~~~~~~~~~
          |                                                                                                                                  |
          |                                                                                                                                  TA_OptInputParameterInfo ** {aka struct TA_OptInputParameterInfo **}
    In file included from talib/_ta_lib.c:615:
    /usr/local/include/ta-lib/ta_abstract.h:446:74: note: expected ‘const TA_OptInputParameterInfo **’ {aka ‘const struct TA_OptInputParameterInfo **’} but argument is of type ‘TA_OptInputParameterInfo **’ {aka ‘struct TA_OptInputParameterInfo **’}
      446 |                                         const TA_OptInputParameterInfo **info );
          |                                         ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~
    talib/_ta_lib.c: In function ‘__pyx_pf_5talib_7_ta_lib_354_ta_getOutputParameterInfo’:
    talib/_ta_lib.c:39007:128: warning: passing argument 3 of ‘TA_GetOutputParameterInfo’ from incompatible pointer type [-Wincompatible-pointer-types]
    39007 |   __pyx_v_retCode = TA_GetOutputParameterInfo(__pyx_f_5talib_7_ta_lib___ta_getFuncHandle(__pyx_v_function_name), __pyx_v_idx, (&__pyx_v_info));
          |                                                                                                                               ~^~~~~~~~~~~~~~
          |                                                                                                                                |
          |                                                                                                                                TA_OutputParameterInfo ** {aka struct TA_OutputParameterInfo **}
    In file included from talib/_ta_lib.c:615:
    /usr/local/include/ta-lib/ta_abstract.h:450:70: note: expected ‘const TA_OutputParameterInfo **’ {aka ‘const struct TA_OutputParameterInfo **’} but argument is of type ‘TA_OutputParameterInfo **’ {aka ‘struct TA_OutputParameterInfo **’}
      450 |                                       const TA_OutputParameterInfo **info );
          |                                       ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~
    talib/_ta_lib.c: In function ‘__pyx_f_5talib_7_ta_lib___ta_getFuncHandle’:
    talib/_ta_lib.c:39606:126: warning: passing argument 2 of ‘TA_GetFuncHandle’ from incompatible pointer type [-Wincompatible-pointer-types]
    39606 |   __pyx_t_1 = __pyx_f_5talib_7_ta_lib__ta_check_success(__pyx_n_s_TA_GetFuncHandle, TA_GetFuncHandle(__pyx_v_function_name, (&__pyx_v_handle)), 0); if (unlikely(!__pyx_t_1)) __PYX_ERR(1, 720, __pyx_L1_error)
          |                                                                                                                             ~^~~~~~~~~~~~~~~~
          |                                                                                                                              |
          |                                                                                                                              TA_FuncHandle ** {aka unsigned int **}
    In file included from talib/_ta_lib.c:615:
    /usr/local/include/ta-lib/ta_abstract.h:155:52: note: expected ‘const TA_FuncHandle **’ {aka ‘const unsigned int **’} but argument is of type ‘TA_FuncHandle **’ {aka ‘unsigned int **’}
      155 |                              const TA_FuncHandle **handle );
          |                              ~~~~~~~~~~~~~~~~~~~~~~^~~~~~
    x86_64-linux-gnu-gcc: fatal error: Killed signal terminated program cc1
    compilation terminated.
    error: command 'x86_64-linux-gnu-gcc' failed with exit status 1
    ----------------------------------------
ERROR: Command errored out with exit status 1: /home/ubuntu/freqtrade/.env/bin/python3.8 -u -c 'import io, os, sys, setuptools, tokenize; sys.argv[0] = '"'"'/tmp/pip-install-pi3sc42f/ta-lib_a20f0cc09394411686db806a1834e6eb/setup.py'"'"'; __file__='"'"'/tmp/pip-install-pi3sc42f/ta-lib_a20f0cc09394411686db806a1834e6eb/setup.py'"'"';f = getattr(tokenize, '"'"'open'"'"', open)(__file__) if os.path.exists(__file__) else io.StringIO('"'"'from setuptools import setup; setup()'"'"');code = f.read().replace('"'"'\r\n'"'"', '"'"'\n'"'"');f.close();exec(compile(code, __file__, '"'"'exec'"'"'))' install --record /tmp/pip-record-k84bu4gk/install-record.txt --single-version-externally-managed --compile --install-headers /home/ubuntu/freqtrade/.env/include/site/python3.8/TA-Lib Check the logs for full command output.
Failed installing dependencies
```

Reply:
yeah - it's receiving a kill-signal by the OS (system-level logs should show a "sacrificed one offending process" - or however the wording is)